sdk/compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/Types.hs

```diff
@@ -89,16 +103,24 @@ instance Ord SubIDEInstance where
 -- We store an optional main ide, the currently closing ides (kept only so they can reply to their shutdowns), and open files
 -- open files must outlive the main subide so we can re-send the TextDocumentDidOpen messages on new ide startup
 data SubIDEData = SubIDEData
-  { ideDataMain :: Maybe SubIDEInstance
+  { ideDataHome :: FilePath
+  , ideDataMain :: Maybe SubIDEInstance
   , ideDataClosing :: Set.Set SubIDEInstance
   , ideDataOpenFiles :: Set.Set FilePath
+  , ideDataFailTimes :: [UTCTime]
+  , ideDataDisabled :: Bool
+  , ideDataLastError :: Maybe String
   }
 
-defaultSubIDEData :: SubIDEData
-defaultSubIDEData = SubIDEData Nothing Set.empty Set.empty
+defaultSubIDEData :: FilePath -> SubIDEData
+defaultSubIDEData home = SubIDEData home Nothing Set.empty Set.empty [] False Nothing
 
 lookupSubIde :: FilePath -> SubIDEs -> SubIDEData
-lookupSubIde home ides = fromMaybe defaultSubIDEData $ Map.lookup home ides
+lookupSubIde home ides = fromMaybe (defaultSubIDEData home) $ Map.lookup home ides
+
+ideShouldDisable :: SubIDEData -> Bool
+ideShouldDisable (ideDataFailTimes -> [t1, t2]) = t1 `diffUTCTime` t2 < 5
```

Comment:
A few comments:
1. This won't work if the subIDE fails more than twice, since it matches only if the list `ideDataFailTimes` has length == 2; the pattern should be something like `(take 2 . ideDataFailTimes -> [t1, t2])`. I see that `ideDataFailTimes` is only ever set to an empty list or an explicit `take 2` expression but that's a tough invariant to keep, imo.
2. The message `SubIDE failed twice within 5 seconds, disabling SubIDE` is hardcoded in `MultiIde.hs` while the 5 seconds are hardcoded here. Perhaps it should be an argument to this function so the message and the delay are specified in the same place. We might even want to make the delay configurable at some point(?)
3. I'm reminded of the ghcide (or is it vscode?) behaviour where if the language server fails five times in close succession it is not restarted again. Maybe there are cases where two failures in 5 seconds does not mean we can't recover?

Reply:
I've implemented the first 2 points, I don't think I want to change the last one though. The delay is already pretty bad on Windows.